tests/basic/execution.rs

```diff
@@ -169,3 +160,116 @@ fn max_steps_early_exit_scheduler() {
 
     assert_eq!(counter_clone.load(Ordering::SeqCst), 50 * 10);
 }
+
+#[test]
+#[should_panic]
+fn context_switches_outside_execution() {
+    shuttle::context_switches();
+}
+
+#[test]
+fn context_switches_atomic() {
+    check_random(
+        move || {
+            let mut threads = vec![];
+            // The actual atomic we test
+            let counter = Arc::new(AtomicUsize::new(0));
+            // Specification counters
```

Comment:
These feel too closely tied to the implementation and so maybe not worth testing like this — I think I’d rather just put the math in a comment (like we do in the DFS tests).

Reply:
Yeah, unfortunately doing something precise here is somewhat tied to the implementation. I definitely wanted to have an assertion in the threads (during the execution), not just at the end. What I could do is to get rid of the std counters and instead do this in the threads:

```rust
let count = counter.fetch_add(1, Ordering::SeqCst) + 1;

// We saw at least one context switch for each `fetch_add`.
assert!(shuttle::context_switches() >= count);

// We did not see the last context switch of this thread.
assert!(shuttle::context_switches() < 11);
```

The first assertion could be strengthened to `>= 2 + 2 * count` in the current implementation (initial context switch, spawn and first context switch per `fetch_add`, and second context switch for `fetch_add` of current thread). It would be more work to update the test when we optimize context switches, but if we assert the final expected number of context switches at the end, we would anyway have to update the test in that case. How much precision do you want?

Reply:
I like tying the assertion to the current value of the counter like that. I think strengthening it to `>= 2 + 2 * count` is also good as long as there's a comment explaining why that's the magic formula.